### ISSUE TITLE
Reindex speed on 8 milion records

### ISSUE THREAD

**Issue by johanpiet2** [artefactual/atom#993]
This two functions slows it down to 0.85ms. I changed it and it does 0.09ms

arElasticSearchInformationObjectPdo.class.php

  private function getCollectionRootId() //slows down to 0.224ms
  {
    //$sql = 'SELECT id FROM information_object force index (information_object_FI_17) 
    //  WHERE parent_id = ? AND lft < ? AND rgt > ?
    //';
	$sql = 'SELECT id FROM (SELECT * FROM information_object WHERE lft < ? AND rgt > ?  ) a WHERE a.parent_id = ?';

  return QubitPdo::fetchColumn($sql, array(QubitInformationObject::ROOT_ID, $this->lft, $this->rgt));

  public function getAncestors()
  {
    if (!isset($this->ancestors))
    {
      // Find ancestors
      $sql  = 'SELECT
                  node.id,
                  identifier,
                  repository_id,
                  slug';
      $sql .= ' FROM '.QubitInformationObject::TABLE_NAME.' node force index (information_object_FI_18)';
      $sql .= ' JOIN '.QubitSlug::TABLE_NAME.' slug
                  ON node.id = slug.object_id';
      $sql .= ' WHERE node.lft < ? AND node.rgt > ?';
      $sql .= ' ORDER BY lft';

      $this->ancestors = QubitPdo::fetchAll($sql, array($this->__get('lft'), $this->__get('rgt')));

information_object_FI_18 = index on lft and rgt

**Comment by jraddaoui** [comment]
Hi @johanpiet2!

I think we talked a bit about this in the use forum. but I'm sorry we didn't address this issue. After one year, and thanks to the upgrade to MySQL 8, this query has been changed to use recursive CTE, which improves the performance even further. There is more information in this Redmine ticket:

https://projects.artefactual.com/issues/13224

We also added several improvements in the indexing process for 2.5.4, and specially for 2.6 in:

https://projects.artefactual.com/issues/13261

Again, thanks for the suggestion.

**Comment by johanpiet2** [comment]
Thank you I will have a look-------- Original message --------From: José Raddaoui Marín <notifications@github.com>Date: Fri, Nov 20, 2020, 17:06To: artefactual/atom <atom@noreply.github.com>Cc: Johan Pieterse <pieterse.johan3@gmail.com>, Mention <mention@noreply.github.com>Subject: Re: [artefactual/atom] Reindex speed on 8 milion records (#993)
Hi @johanpiet2!
I think we talked a bit about this in the use forum. but I'm sorry we didn't address this issue. After one year, and thanks to the upgrade to MySQL 8, this query has been changed to use recursive CTE, which improves the performance even further. There is more information in this Redmine ticket:
https://projects.artefactual.com/issues/13224
We also added several improvements in the indexing process for 2.5.4, and specially for 2.6 in:
https://projects.artefactual.com/issues/13261
Again, thanks for the suggestion.
